pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1803,6 +1808,77 @@ private boolean updateExistedSegment(SegmentZKMetadata segmentZKMetadata) {
     return instancesToSegmentsMap;
   }
 
+  public Map<String, String> getSegmentsCrcForTable(String tableName) {
+    // Get the segment list for this table
+    IdealState is = _helixAdmin.getResourceIdealState(_helixClusterName, tableName);
+    List<String> segments = new ArrayList<>(is.getPartitionSet());
+
+    // Make a list of segment metadata for the given table
+    List<String> segmentMetadataPaths = new ArrayList<>(segments.size());
+    for (String segmentName : segments) {
+      segmentMetadataPaths.add(buildPathForSegmentMetadata(tableName, segmentName));
+    }
+
+    // Initialize cache if it is the first time to process the table.
+    if (!_lastKnownSegmentMetadataVersionMap.containsKey(tableName)) {
+      _lastKnownSegmentMetadataVersionMap.put(tableName, new ConcurrentHashMap<String, Integer>());
+    }
+    if (!_segmentCrcMap.containsKey(tableName)) {
+      _segmentCrcMap.put(tableName, new ConcurrentHashMap<String, Long>());
+    }
+
+    // Get ZNode stats for all segment metadata
+    Stat[] metadataStats = _propertyStore.getStats(segmentMetadataPaths, AccessOption.PERSISTENT);
+
+    // Update the crc information for segments that are updated
+    for (int i = 0; i < metadataStats.length; i++) {
+      String currentSegment = segments.get(i);
+      Stat metadataStat = metadataStats[i];
+      if (metadataStat != null) {
```

Comment:
If the segment zk metadata for the segment is somehow missing, this can be null.